worker/deployer/manifold.go

```diff
@@ -37,16 +46,51 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
 // It's not tested at the moment, because the scaffolding
 // necessary is too unwieldy/distracting to introduce at this point.
 func (config ManifoldConfig) newWorker(a agent.Agent, apiCaller base.APICaller) (worker.Worker, error) {
+	// TODO: run config.Validate()
 	cfg := a.CurrentConfig()
 	// Grab the tag and ensure that it's for a machine.
 	if cfg.Tag().Kind() != names.MachineTagKind {
 		return nil, errors.New("agent's tag is not a machine tag")
 	}
 	deployerFacade := apideployer.NewState(apiCaller)
-	context := config.NewDeployContext(deployerFacade, cfg)
-	w, err := NewDeployer(deployerFacade, context)
+	contextConfig := ContextConfig{
+		Agent:            a,
+		Clock:            config.Clock,
+		Logger:           config.Logger,
+		UnitEngineConfig: config.UnitEngineConfig,
+		SetupLogging:     config.SetupLogging,
+		UnitManifolds:    UnitManifolds,
+	}
+
+	context, err := config.NewDeployContext(contextConfig)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	shim := &apiShim{deployerFacade}
+	w, err := NewDeployer(shim, config.Logger, context)
 	if err != nil {
 		return nil, errors.Annotate(err, "cannot start unit agent deployer worker")
 	}
 	return w, nil
 }
+
+type apiShim struct {
+	st *apideployer.State
+}
+
+func (s *apiShim) Machine(tag names.MachineTag) (Machine, error) {
+	// Need to deal with typed nils.
+	machine, err := s.st.Machine(tag)
```

Comment:
This can also get simplified down to `return s.st.Machine(tag)`

Reply:
Unfortunately it can't.

I hit this problem with the description package as well.

`s.st.Machine(tag)` returns a *state.Machine, and if the returned value is nil, it is a typed nil, meaning that the type of the interface is a *state.Machine and the value is nil, which means when you compare against nil with `shim.Machine(tag) == nil` the typed nil doesn't match. This is a trick for new players.